### PR TITLE
Move GPIO Pad Control for BCM2835 page from scribd to GitHub

### PR DESCRIPTION
**Pull request by rbisewski**
As part of issue #971, some of the GPIO pad control documentation was located on scribd, which made it hard to browse or edit.

I have since reproduced that document, including table and image, and converted it into a markdown document in the Raspberry Pi documentation GitHub repo.

Let me know what you think and feel free to suggest and edits.

### CODE REVIEW

File: hardware/raspberrypi/gpio/gpio_pads_control2.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# GPIO pads control
+
+Extracted from BCM2835 full data sheet, from Gert van Loo 2-August-2012. Rewritten into this markdown document by Robert Bisewski in 21-October-2018.
+
```
Comment (JamesH65) [inline hardware/raspberrypi/gpio/gpio_pads_control2.md]: We don't have any attribution (except CC stuff at bottom of pages) in our documentation so would prefer this section be removed.

File: hardware/raspberrypi/gpio/gpio_pads_control2.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+### Errata
+
+In the previous version on scribd, I accidentally dropped the password bits (bits 31..24) which need to be 5A in order to write a different value to the register!
+
```
Comment (JamesH65) [inline hardware/raspberrypi/gpio/gpio_pads_control2.md]: Does this errata section need to be here? Should be included already in the body below.

**Comment by JamesH65** [comment]
Hi @rbisewski, thanks very much for contributing to our docs repo - nice to get some third party contributions! 

I've added some review comments on the PR's you have submitted, although have left the NFS one along as this still appears to be a work in progress.

**Comment by rbisewski** [comment]
No problem, thanks for reviewing my changes.

I have since removed the sections you requested, feel free to merge once you're satisfied.

**Comment by bennuttall** [comment]
Why is the new file named with a 2?

**Comment by rbisewski** [comment]
@bennuttall mostly because the old scribd file had that as part of its name.

I suppose there is no reason to call it that anymore. I have since removed the 2 from the name.

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+1. It is **not**: the current that the pad will deliver.
+1. It is **not**: a current limit so the pad will not blow up.
+
+The pad output is a voltage source:
+
+* if set high the pad will try to drive the output to the rail voltage which on the Raspberry-Pi is 3V3 (3.3 Volts)
+* if set low the pad will try to drive the output to ground (0 Volts)
+
+As the text says: the pad will try to drive the output high or low. If it succeeds depends on what is connected. If the pas is shorted to ground it will not be able to drive high. In fact itwill try to deliver as much current as it can and the current is only limited to what the internal resistance is.
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: pas -> pad
itwill -> it will

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+#### Why don't I set all my pads to the maximum current?
+
+Two reasons:
+
+1. The raspberry-Pi 3V3 supply was designed with a maximum current of ~3mA per GPIO pin. If you load each pin with 16mA the total current is 272mA. The 3V3 supply will collapse under that!
+1. Big current spikes will happen especially if you have a capacitive load. That will "bounce" around all the other pins near it. It is likely to cause interference with the SD-card or even the SDRAM behaviour.
+
+#### What is a safe current?
+
+All the electronics of the pads are designed for 16mA. That is a safe value under which youwill not damage the device. Even if you set the drive strength to 2mA and then load it so 16mA comes out that will not damage the device. Other than that there is no guaranteed maximum safe current.
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: youwill -> you will

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+   
+Thus a drive strength of 16mA means:
+
+If you set the pad high you can draw up to 16mA and we still guarantee that the output voltage will be >=1.3V. This also means that if you set a drive strength of 2mA and you draw 16mA the voltage will **not** be 1.3 Volt but lower. In fact it may not be high enough to be seen as high by an external device.
+
+#### Why don't I set all my pads to the maximum current?
+
+Two reasons:
+
+1. The raspberry-Pi 3V3 supply was designed with a maximum current of ~3mA per GPIO pin. If you load each pin with 16mA the total current is 272mA. The 3V3 supply will collapse under that!
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: raspberry-Pi -> Raspberry Pi

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 6]
```diff
@@ -1,41 +1,10 @@
 # GPIO pads control
 
-### Address
+This document expains the current capabilities of the Raspberry Pi GPIO pins. It is applicable to all models up to an including the Mode 3B+. The main thing to note is that the GPIO drive strengths do not indicate a maximum current, but a maximum current under which the pad will still meet the specification. You should set the GPIO drive strengths to match the device being attached in order for the device to work correctly.
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: expains -> explains
an including -> and including
Mode 3B+ -> Model 3B+

:slightly_smiling_face:

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 7]
```diff
@@ -47,40 +16,64 @@ still meet the specification*.
...
-
-This has to do with the voltage levels guaranteed. As this is a digital pad there are two voltage levels: high and low. Yet how high and how low? Is high 0.9V or 1.5V or 2.7V or what?
+Coming back to the statement that the current value specifies the maximum current under which the pad will still meet the specification.
+This has to do with the guaranteed voltage levels. As this is a digital pad there are two voltage levels, high and low. Yet how high and how low? Is high 0.9V or 1.5V or 2.7V or some other voltage?
 
 To answer that question I/O ports have two parameters which deal with the output level:
-*  V<sub>IL</sub>: The maximum low level voltage. (0.8V on the BCM2835)
-*  V<sub>IH</sub> : The minimum high level voltage. (1.3V on the BCM2835)
+*  V<sub>IL</sub>: The maximum low level voltage. (0.8V on the BCM2835/6/7)
+*  V<sub>IH</sub> : The minimum high level voltage. (1.3V on the BCM28356/7)
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: missing slash between 5 and 6

**Comment by JamesH65** [comment]
Made a couple of final polish changes, will now push to copy editors.

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 8]
```diff
@@ -10,27 +10,24 @@ Inside the pad there are a number of drivers, all in parallel. If the drive stre
...
+Meeting the specification is determined by the guaranteed voltage levels. Because the pads are digital, there are two voltage levels, high and low. The I/O ports have two parameters which deal with the output level:
 
-To answer that question I/O ports have two parameters which deal with the output level:
-*  V<sub>IL</sub>: The maximum low level voltage. (0.8V on the BCM2835/6/7)
-*  V<sub>IH</sub> : The minimum high level voltage. (1.3V on the BCM2835/6/7)
+*  V<sub>IL</sub>: The maximum low level voltage. (0.8V)
+*  V<sub>IH</sub> : The minimum high level voltage. (1.3V)
 
 V<sub>IL</sub>=0.8V means that if the output is Low it will be <= 0.8V.
 V<sub>IL</sub>=1.3V means that if the output is High it will be >= 1.3V.
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: I guess this should be `V<sub>IH</sub>=1.3V` ?

**Comment by lurch** [comment]
I just noticed that the table at https://github.com/raspberrypi/documentation/blob/master/hardware/raspberrypi/gpio/README.md#voltage-specifications lists V(IL) of 0.8V and V(IH) of 1.3V (which matches up with the descriptions in this pad-control PR) when VDD IO = 2.7V. But don't all models of Raspberry Pi (ignoring the ComputeModules) have VDD IO = 3.3V ?  :man_shrugging:

**Comment by JamesH65** [comment]
I'll need to cross check with Roger/Mike on that one, was wondering the same thing myself.

File: hardware/raspberrypi/gpio/gpio_pads_control.md [hunk 9]
```diff
@@ -26,17 +26,17 @@ If the pad is driven high and it is shorted to ground in due time it will fail.
...
 
-V<sub>IL</sub>=0.8V means that if the output is Low it will be <= 0.8V.
-V<sub>IH</sub>=1.3V means that if the output is High it will be >= 1.3V.
+V<sub>IL</sub>=0.9V means that if the output is Low it will be <= 0.9V.
+V<sub>IH</sub>=1.6V means that if the output is High it will be >= 1.6V.
    
 Thus a drive strength of 16mA means:
 
-If you set the pad high you can draw up to 16mA and it is guaranteed that the output voltage will be >=1.3V. This also means that if you set a drive strength of 2mA and you draw 16mA the voltage will **not** be 1.3 Volt but lower. In fact it may not be high enough to be seen as high by an external device.
+If you set the pad high you can draw up to 16mA and it is guaranteed that the output voltage will be >=V<sub>IL</sub>. This also means that if you set a drive strength of 2mA and you draw 16mA the voltage will **not** be V<sub>IL</sub> but lower. In fact it may not be high enough to be seen as high by an external device.
```
Comment (lurch) [inline hardware/raspberrypi/gpio/gpio_pads_control.md]: I suspect you meant to use VIH here rather than VIL?

Comment (JamesH65) [reply hardware/raspberrypi/gpio/gpio_pads_control.md]: Bugger, I thought I had double checked that as well! Quite right. C&P'ed the wrong one.

**Pull request merged**